Add libseat support (run Xserver as ordinary user)

cepelinas9000 opened the pull request:
Adds libseat support, it is adapted devuan patches for Xlibre. 

Firstly, currently draft mode - technically it is nearly verbatium patch (with Xlibre adaptations) from https://git.devuan.org/devuan/xorg-server/compare/suites/experimental...support-libseat-upstream  

Secondly,  with extra work and additional patches: it should be possible to start xserver from AppImage/chroot  - as long there are working device enumeration (or nodes configured beforehand -  seatd opens device nodes for xserver and sents via unix socket as fd's)
 
@b-aaz there are mentioned at https://sr.ht/~kennylevinsen/seatd/ FreeBSD support it working or in early stages (I didn't check deeply) ? 

Fixes: #202

To test it, you need libseat and seatd running (doesn't matter as service - you can start as root via tmux). Only problem, that if you have already running X server libseatd will try revoke display driver node and fails (you need initialize with X running).

If you don't want to install seatd:

```console
$ git clone https://git.sr.ht/~kennylevinsen/seatd
$ cd seatd
$ meson setup --prefix=/tmp/seat build
$ (cd build;ninja install)
....
```

then just start if
```console
# cd /tmp/seatd/bin
localhost /tmp/seatd/bin # ./seatd -u as -g as 
### "-u" and "-g" switches change socket uid/gid for easier testing
```

but don't forget when compiling Xlibre you need to provide seatd directory:
```
PKG_CONFIG_PATH=/tmp/x11libre/lib64/pkgconfig meson setup --prefix=/tmp/x11libre build
```

and when running LD_LIBRARY_PATH
```
$ LD_LIBRARY_PATH=/tmp/seatd/lib64/ ./X :1 -retro -seat seat0 -keeptty vt7
```

these parameters necessary for now (that why draft):
```
 -seat seat0 -keeptty vt7
```

*edit: added test instructions*

cepelinas9000 commented:
> @cepelinas9000 Did you have a look at [comment 3220492682 · Add seatd/libseat support · Issue #202 · X11Libre/xserver](https://github.com/X11Libre/xserver/issues/202#issuecomment-3220492682) and especially [LeePen/xlibre - xlibre - Devuan git store](https://git.devuan.org/LeePen/xlibre)?

Now I have looked :) Missed that comment, Mark have fixed additional blocking bugs (in this PR Xserver refuses to start when there are no '-seat' argument given)

callmetango commented:
> Now I have looked :)

Excellent. :wink: I added a "Fixes: ..." line to the description above so the feature request ticket gets a link to here.

cepelinas9000 commented on hw/xfree86/common/xf86Init.c:
```diff
@@ -848,6 +851,7 @@ ddxGiveUp(enum ExitCode error)
     if (xorgHWOpenConsole)
         xf86CloseConsole();
 
+    seatd_libseat_fini();
```
Yes, the CI tests where passing. The original patch have `#ifdef` and following style code (see, [seatd-libseat.h](https://github.com/X11Libre/xserver/pull/935/files#diff-69cbd8fce417feb4bd845b32f1883791fdf9f0952a361aecb05105723e351786):
```
#ifdef SEATD_LIBSEAT
...
#else
#define seatd_libseat_init()
		
#define seatd_libseat_fini()
#endif
```

cepelinas9000 commented:
> The terminology "rootless" is ambiguous here, because it usually refers to an Xserver without an actual root window.

I changed PR title to less confusing

cepelinas9000 commented:
> @cepelinas9000 I think the target branch of this pull request should be changed from `maint-25.0` to `master` since this is a new feature, no?

I change this to master when will be ready. Now it temporary measure to not deal with intermediate merge conflicts🤞.

b-aaz commented:
Yes it is available on FreeBSD.
https://www.freshports.org/sysutils/seatd/

b-aaz commented:
This is a large change, In my opinion it needs as large of a change in documentation too, things like new flags and configs should be noted.
It would be nice to also include some operational info too, Things like how to setup and run X with seatd.

Another thing that I forgot to mention is that FreeBSD has seatd but elogind is NOT available.

stefan11111 commented on include/seatd-libseat.h:
```diff
@@ -0,0 +1,49 @@
+extern int seatd_libseat_switch_session(int session);
+extern Bool seatd_libseat_controls_session(void);
+#else
+#define seatd_libseat_init()
```
Shouldn't this also become an int?

Do macro stubs like this really work?

We never intend to do something like `&seatd_libseat_init` or `seatd_libseat_switch_session(i++)`?

cepelinas9000 replied:
On other hand to be on absolutely safe side, i changed to inline functions.

callmetango commented:
@cepelinas9000
> Totally don't mind it would be even helpful.

I invited the Devuan devs by posting to [libera/#devuan-dev/ Thursday, 2025-09-18](http://reisenweber.net/irclogs/libera/_devuan-dev/_devuan-dev.2025-09-18.log.html#t2025-09-18T18:34:21).

ONykyf commented:
@cepelinas9000 
> To test it, you need libseat and seatd running (doesn't matter as service - you can start as root via tmux). Only problem, that if you have already running X server libseatd will try revoke display driver node and fails (you need initialize with X running).

I don't know anything about libseat/seatd, but would like to try it. Some questions on the topic:

- libseat/seatd must be initialized _with_ or _without_ X server running?

- can libseat/seatd be restricted to handle only _part_ of hardware, maybe all not initialized yet?

- if yes, is it possible to run several seatd instances at once, or to have one X server started through seatd and another one started in another way?

cepelinas9000 commented:
> I don't know anything about libseat/seatd, but would like to try it. Some questions on the topic:
> 
>     * libseat/seatd must be initialized _with_ or _without_ X server running?
> 
>     * can libseat/seatd be restricted to handle only _part_ of hardware, maybe all not initialized yet?
> 
>     * if yes, is it possible to run several seatd instances at once, or to have one X server started through seatd and another one started in another way?

To answer thesese questions, I want to test myself too (time shortage etc...) and familiarize myself.

Libseat doing following:

* seatd initialization is to perform vt terminal switch  (operation which requires root user) - the sad part (same for logind) it gets current active terminal vt from /dev/tty0 (or equivalent) during client initialization. (relavant code: https://github.com/kennylevinsen/seatd/blob/900e359f99818df434389eb3a39ee767f1d84f2c/seatd/seat.c#L213)... yes it is for `vt_bound` session, but which ones are saw it where vt bound (i didn't performed deeply test with `non vt bound session`). As from previous sentence you see from where  `-keeptty` switch came. (Want to start Xserver on other tty... the start script will look like `sudo chvt 8;sleep 0.1;./X :2 -keeptty`

* libseat only knows `seat0` (that current limitation, logind from systemd is advanced - it tries to limit by `udev seat`), in this context `seat0` mean anything what controls current /dev/tty0 switching and library don't touch devices directly. It  its code only reacts to vt changes. In practice doesn't matter if multiple seatd running - you need to avoid running multiple Xservers on same vt

* what `seatd` does: it open devices (Xserver perform scanning as usual, just seatd open device nodes), passes its fds, and disables/enables (during vt switch) with ioctls: `EVIOCREVOKE`, `HIDIOCREVOKE` and `DRM_IOCTL_SET_MASTER/DRM_IOCTL_DROP_MASTER` - in other words  it forcibly disable/enable any input and output controls.

cepelinas9000 commented:
Now if libseat active, the modesetting driver tries to open gpu device node trough `seatd` (if not - performs as usual). The implication is that: user doesn't not need to have any rights to `/dev/dri/card*`, of course other story that `/dev/dri/renderD*` was directly opened when trying with `strace glxgears`. :

<img width="710" height="188" alt="as_user" src="https://github.com/user-attachments/assets/86ad93a6-58b2-443e-a181-3e4904b9cff8" />

Refactored  code should work on freebsd (or any bsd) as long there is `seatd` activate (possible to connect to `/var/run/seatd.sock`) and `kms` video kernel driver. (couldn't test freebsd because qemu gpu driver was not kms).

cepelinas9000 commented on hw/xfree86/drivers/modesetting/driver.c:
```diff
@@ -226,6 +231,28 @@ open_hw(const char *dev)
+#ifdef SEATD_LIBSEAT
+   // try get device from seatd arbiter
+   if (dev){
+     fd = seatd_libseat_open_graphics(dev);
```
@X11Libre/wranglers ping.  @metux @stefan11111 

I took liberty and increased public sdk surface with `seatd_libseat_open_graphics`  export (it asks seatd server to perform open operation and via fd passing returns **file fd**), my question is: shouldn't   (in far *future*) be some sort api* o deal with device opening. Or I missed something?

* for example amdgpu calls open directly, same intel

cepelinas9000 commented:
moved opening device node down - now tries open as current process firstly then tries call seatd and improved failure case message (see: https://github.com/X11Libre/xserver/pull/935/files#diff-54b8e9df3c29f5a215f2ffe63e1d897036c6263c7699a3be0d1bbb53e7e040f6R120 )

stefan11111 commented on hw/xfree86/common/xf86Init.c:
```diff
@@ -433,9 +436,11 @@ InitOutput(ScreenInfo * pScreenInfo, int argc, char **argv)
-            xf86OpenConsole();
-        else
+        if (xorgHWOpenConsole) {
+	    if (!seatd_libseat_controls_session()) {
```
Indentation looks wrong on github here for some reason.
Did you use tabs instead of spaces?

cepelinas9000 replied:
That my mixup didn't notice that original diff had tabs here, i just copied from it.

stefan11111 commented on hw/xfree86/os-support/meson.build:
```diff
@@ -157,6 +158,15 @@ else
     ]
 endif
 
+if host_machine.system() == 'linux' or host_machine.system().endswith('bsd')
```
@b-aaz Is this the correct way to check for bsd?

stefan11111 commented on hw/xfree86/os-support/shared/drm_platform.c:
```diff
@@ -21,34 +21,44 @@
 
-    if (fd == -1)
+    if (fd == -1) {
+        // Try opening the path directly
```
Use a C-style comment

stefan11111 commented on hw/xfree86/os-support/shared/seatd-libseat.c:
```diff
@@ -0,0 +1,400 @@
+            }
+        }
+    xf86InputEnableVTProbe(); // Add any paused input devices
+    xf86platformVTProbe(); // Probe for outputs
```
Use a C-style comment

stefan11111 commented on hw/xfree86/os-support/shared/seatd-libseat.c:
```diff
@@ -0,0 +1,400 @@
+    free(xfmt);
+}
+
+// ============== seatd-libseat.h API functions =============
```
Use a C-style comment

cepelinas9000 commented:
> @cepelinas9000 This pr is intended for maint? Not master?

To master - when it will be +/- ok then i rebase it.

cepelinas9000 commented on hw/xfree86/drivers/modesetting/driver.c:
```diff
@@ -235,6 +242,30 @@ open_hw(const char *dev)
+
+#ifdef SEATD_LIBSEAT
+   // try get device from seatd arbiter
+   if (seatd_libseat_controls_session()){
```
> I'm not familiar with this, but perhaps you can take a look at the logind code?
I don't see any logind-specific code in the drivers.

When everything works™, device is opened with same `seatd_libseat_open_graphics(dev)`, during `get_drm_info` probing ( hw/xfree86/os-support/shared/drm_platform.c#L26 )  in driver with help `xf86_platform_device_odev_attributes` (search in driver.c)  call. 

But if there is some failure, then driver call `hw_open` function and tries manually open device. The systemd-logind code doesn't handle this case at all...

> Also, how exactly does fd = seatd_libseat_open_graphics(dev); work?
You can't just pass fd's from the daemon to the X server.

It work exactly passing fd's: opens fd from elevated process, via unix socket sends fd to xserver. The twist is that it saves fd inside and additionally calls drop master / revoke ioctls (why it works - 'dup semantics here', you can play with example https://gist.github.com/cepelinas9000/8f30a41894adb84c2786b4d658ac73c9 ) 


> Does it open a socket to the seatd process and pass the data to/from the real fd? What about the ioctl's we use on the graphics device?


Yes it opens session during initialization,  for `seatd` unix socket or dbus connection if systemd-logind. The `ioctl` business as usual - it return `fd` and that all. Obviously if seat/logind crashes - worse case scenario during `chvt` you won't return from tty text mode.

metux commented on hw/xfree86/drivers/video/modesetting/driver.c:
```diff
@@ -247,6 +254,31 @@ open_hw(const char *dev)
+
+#ifdef SEATD_LIBSEAT
+    /* try get device from seatd arbiter */
+    if (fd == -1 && seatd_libseat_controls_session()){
```
perhaps move the whole thing into the seatd support code ? (instead of potentially duplicated within drivers)

cepelinas9000 replied:
refactored into direct aproach: just call `seatd_libseat_open_graphics` - if some problems it return <0 (not -1 like open)

cepelinas9000 commented on hw/xfree86/os-support/shared/drm_platform.c:
```diff
@@ -21,34 +21,44 @@
 
     LogMessage(X_INFO, "Platform probe for %s\n", attribs->syspath);
 
-    fd = systemd_logind_take_fd(attribs->major, attribs->minor, path, &paused);
```
I doens't matter: seatd and systemd-elogind are mutually exclusive - only one can be turned on

cepelinas9000 commented on meson.build:
```diff
@@ -110,6 +110,9 @@ xfont2_dep = dependency('xfont2', version: '>= 2.0')
 dbus_required = get_option('systemd_logind') == 'true'
 dbus_dep = dependency('dbus-1', version: '>= 1.0', required: dbus_required)
 
+seat_libseat_required = get_option('seatd_libseat') == 'true'
```
There are little problem here:
* libseat and systemd-elogind are mutually exclusive: you can set only one
* when set `-Dsystemd_logind=auto` and `-Dseatd_libseat=true` - the `systemd_logind` automatically got disabled
* `systemd_logind` depends on `udev_kms` & `dbus` - they need to be moved in somewhere around `hw/xfree86/common` too

I can  make pr's to move `udev_kms` `dbus` `systemd_logind` to  `hw/xfree86/common` or similar place firstly then refactor this pr

cepelinas9000 commented:
I pushed new fix:
* only `seatd_libseat_open_graphics` is _X_EXPORT - function already internally checking everything whats necessary
*  added comment why `seatd_libseat_open_graphics` necessary to export, with question for future `@note XXX: maybe in future Xlibre public api could gain function for opening device nodes by path?`
* fixed warning when building with `-Dsystemd_logind=true`

metux commented on hw/xfree86/os-support/shared/seatd-libseat.c:
```diff
@@ -44,7 +44,7 @@
 #include "seatd-libseat.h"
 
-// ============ libseat client adapter ======================
+/* ============ libseat client adapter ====================== */
```
please squash those into previous commit.

cepelinas9000 commented on hw/xfree86/drivers/video/modesetting/driver.c:
```diff
@@ -240,17 +245,32 @@ open_hw(const char *dev)
+	#ifdef SEATD_LIBSEAT
+	/* try to open dev node via libseat */
+	if (fd == -1) {
+	    fd = seatd_libseat_open_graphics(dev);
```
> I still have my concerns about the fact that we can't just open the device and we need to have libseat open it for us.

With seatd indirection - you can open  `/dev/dri/card*` without any permissions (in ideal world xserver should give fd from `/dev/dri/render*` for clients too - then you could have hw acceleration from docker/podman with blocked access to device nodes [i mean  cgroup devices.allow/devices.deny] )


>  I'm guessing libseat will only work in modesetting?

It more complicated - the happy path is to open device during probe and give to driver (see: `probe_hw` function around `#ifdef XF86_PDEV_SERVER_FD`, similar logic in other drivers - here in xserver core asks (e)logind, seatd and so on for device ).

additionally (e)logind & seatd handes handles input devices - big chunk code in `common/xf86Xinput.c`

 > How does the (e)logind code deal with this? I don't see any (e)logind-specific code here.

it not handling for that case at all -  you will just end with `permission denied for /dev/dri/card*` error, .

cepelinas9000 commented:
@callmetango @stefan11111 @b-aaz @ONykyf  ping

cepelinas9000 commented on hw/xfree86/common/xf86Events.c:
```diff
@@ -225,8 +233,10 @@ xf86ProcessActionEvent(ActionEvent action, void *arg)
-    if (xf86VTSwitchPending())
-        xf86VTSwitch();
+    if (xf86VTSwitchPending() ||
+        (dispatchException & DE_TERMINATE)){
```
it was in devuan patch `https://git.devuan.org/devuan/xorg-server/commit/372bb886e01b6c59fd19cb2c1703567b5deedc25`. It looks original motivation is to start to drop drm master sooner to make faster exit.

cepelinas9000 replied:
> You mean vt master?

With this code parts  I'm not too familiar, but skimming though code it looked like this motivation. Or i'm wrong, @metux ?

metux replied:
ok. then give me your ACKs, and I'll merge it.
